src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs

```diff
@@ -11,18 +12,27 @@ namespace System.Collections.Generic
 {
     #region ArraySortHelper for single arrays
 
-    internal partial class ArraySortHelper<T>
+    internal partial class ArraySortHelper<T, TComparer>
+        where TComparer : IComparer<T>?
     {
         #region IArraySortHelper<T> Members
 
-        public void Sort(Span<T> keys, IComparer<T>? comparer)
+        public void Sort(Span<T> keys, TComparer comparer)
         {
             // Add a try block here to detect IComparers (or their
             // underlying IComparables, etc) that are bogus.
             try
             {
-                comparer ??= Comparer<T>.Default;
-                IntrospectiveSort(keys, comparer.Compare);
+                if (comparer is null)
```

Comment:
This will create two instantiations of the sorting code: One on `Comparer<T>` and second on `TComparer`.

Can the null check be pushed out to the callers to avoid the duplication where possible?

Reply:
@jkotas what if we put `!typeof(TComparer).IsValueType && comparer is null`?

Reply:
`comparer is null` is JITed into a constant already for ValueTypes. I do not think this would help.

Reply:
yeah I thought so, perhaps I don't fully understand your issue here then given only reference type TComparer should be an issue but that should have a canonical instantiation or? would you mind expanding?

Reply:
For example, when this is called from here:

     https://github.com/dotnet/runtime/pull/39543/files#diff-d4e4a789c4e124d267dde2cf6505da8eR1760

`comparer` will be `null` and so we will always take the first branch (as long as this is the only `Sort` use for the given `T`). The JIT or AOT won't be able to figure it out. They will create both generic instantiations of the sorting algorithm.

> a canonical instantiation

Yes, instantiations over reference types share code, but there is still duplication of the type system structures.

Reply:
> there is still duplication of the type system structures.

Right, of course :) Just ball-parking here, but would the following then not ensure we only have a single type system structure for reference type TComparer?
```csharp
                if (typeof(TComparer).IsValueType)
                {
                    ComparerArraySortHelper<TKey, TValue, TComparer>
                        .IntrospectiveSort(keys, values, comparer);
                }
                else 
                {
                    IComparer<TKey> referenceComparer = comparer ?? Comparer<TKey>.Default;
                    ComparerArraySortHelper<TKey, TValue, IComparer<TKey>>
                        .IntrospectiveSort(keys, values, referenceComparer);
                }
```